### ISSUE TITLE
Cluster: Can't LXC list, DQLite can't connect, "some nodes are behind this node's version"

### ISSUE THREAD

**Issue by qskousen** [lxc/lxd#7540]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04
 * The output of "lxc info" or if that fails: (it fails)
   * Kernel version: 4.15.0-66-generic
   * LXC version: 4.2
   * LXD version: 4.2 (snap stable)
   * Storage backend in use: ZFS

# Issue description

Cluster can't connect to eachother on DQLite. All LXC commands hang forever.
Log shows:

```
Jun 16 13:01:06 monkeyweb1 lxd.daemon[6539]: t=2020-06-16T13:01:06-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.159.233:8443: dial: 503 Service Unavailable"
Jun 16 13:01:06 monkeyweb1 lxd.daemon[6539]: t=2020-06-16T13:01:06-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.210.116:8443: dial: 503 Service Unavailable"
Jun 16 13:01:06 monkeyweb1 lxd.daemon[6539]: t=2020-06-16T13:01:06-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.211.119:8443: dial: 503 Service Unavailable"
Jun 16 13:01:06 monkeyweb1 lxd.daemon[6539]: t=2020-06-16T13:01:06-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.220.80:8443: dial: some nodes are behind this node's version"
```

# Steps to reproduce

Unknown.

# Information to attach

I have run `sqlite3 /var/snap/lxd/common/lxd/database/local.db .dump` on all nodes and they all seem to agree except one has in `sqlite_sequence` `'config', 3` where the others have 2, but there is no extra config entry, the last one is numbered 3 instead of 2:

```
$ sudo sqlite3 /var/snap/lxd/common/lxd/database/local.db .dump
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE schema (
    id         INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
    version    INTEGER NOT NULL,
    updated_at DATETIME NOT NULL,
    UNIQUE (version)
);
INSERT INTO schema VALUES(1,38,1557965180);
INSERT INTO schema VALUES(2,39,1580760435);
CREATE TABLE config (
    id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
    key VARCHAR(255) NOT NULL,
    value TEXT,
    UNIQUE (key)
);
INSERT INTO config VALUES(1,'cluster.https_address','192.168.211.119:8443');
INSERT INTO config VALUES(3,'core.https_address','xx.xx.xxx.xxx');
CREATE TABLE patches (
    id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
    name VARCHAR(255) NOT NULL,
    applied_at DATETIME NOT NULL,
    UNIQUE (name)
);
INSERT INTO patches VALUES(1,'shrink_logs_db_file',1557965180);
INSERT INTO patches VALUES(2,'invalid_profile_names',1557965180);
INSERT INTO patches VALUES(3,'leftover_profile_config',1557965180);
INSERT INTO patches VALUES(4,'network_permissions',1557965180);
INSERT INTO patches VALUES(5,'storage_api',1557965180);
INSERT INTO patches VALUES(6,'storage_api_v1',1557965180);
INSERT INTO patches VALUES(7,'storage_api_dir_cleanup',1557965180);
INSERT INTO patches VALUES(8,'storage_api_lvm_keys',1557965180);
INSERT INTO patches VALUES(9,'storage_api_keys',1557965180);
INSERT INTO patches VALUES(10,'storage_api_update_storage_configs',1557965180);
INSERT INTO patches VALUES(11,'storage_api_lxd_on_btrfs',1557965180);
INSERT INTO patches VALUES(12,'storage_api_lvm_detect_lv_size',1557965180);
INSERT INTO patches VALUES(13,'storage_api_insert_zfs_driver',1557965180);
INSERT INTO patches VALUES(14,'storage_zfs_noauto',1557965180);
INSERT INTO patches VALUES(15,'storage_zfs_volume_size',1557965180);
INSERT INTO patches VALUES(16,'network_dnsmasq_hosts',1557965180);
INSERT INTO patches VALUES(17,'storage_api_dir_bind_mount',1557965180);
INSERT INTO patches VALUES(18,'fix_uploaded_at',1557965180);
INSERT INTO patches VALUES(19,'storage_api_ceph_size_remove',1557965180);
INSERT INTO patches VALUES(20,'devices_new_naming_scheme',1557965180);
INSERT INTO patches VALUES(21,'storage_api_permissions',1557965180);
INSERT INTO patches VALUES(22,'container_config_regen',1557965180);
INSERT INTO patches VALUES(23,'lvm_node_specific_config_keys',1557965180);
INSERT INTO patches VALUES(24,'candid_rename_config_key',1557965180);
INSERT INTO patches VALUES(25,'move_backups',1557965180);
INSERT INTO patches VALUES(26,'storage_api_rename_container_snapshots_dir',1557965180);
INSERT INTO patches VALUES(27,'storage_api_rename_container_snapshots_links',1557965180);
INSERT INTO patches VALUES(28,'fix_lvm_pool_volume_names',1557965180);
INSERT INTO patches VALUES(29,'storage_api_rename_container_snapshots_dir_again',1560825397);
INSERT INTO patches VALUES(30,'storage_api_rename_container_snapshots_links_again',1560825397);
INSERT INTO patches VALUES(31,'storage_api_rename_container_snapshots_dir_again_again',1560998190);
INSERT INTO patches VALUES(32,'clustering_add_roles',1571939017);
INSERT INTO patches VALUES(33,'clustering_add_roles_again',1575331415);
INSERT INTO patches VALUES(34,'storage_create_vm',1579796978);
INSERT INTO patches VALUES(35,'storage_zfs_mount',1579988334);
INSERT INTO patches VALUES(36,'network_pid_files',1584432997);
INSERT INTO patches VALUES(37,'storage_create_vm_again',1584432997);
INSERT INTO patches VALUES(38,'storage_zfs_volmode',1584432997);
INSERT INTO patches VALUES(39,'storage_rename_custom_volume_add_project',1584991587);
INSERT INTO patches VALUES(40,'storage_lvm_skipactivation',1590426018);
CREATE TABLE raft_nodes (
    id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
    address TEXT NOT NULL, role INTEGER NOT NULL DEFAULT 0,
    UNIQUE (address)
);
INSERT INTO raft_nodes VALUES(1,'192.168.211.119:8443',0);
INSERT INTO raft_nodes VALUES(3,'192.168.159.233:8443',2);
INSERT INTO raft_nodes VALUES(4,'192.168.220.80:8443',2);
INSERT INTO raft_nodes VALUES(5,'192.168.210.116:8443',0);
DELETE FROM sqlite_sequence;
INSERT INTO sqlite_sequence VALUES('schema',2);
INSERT INTO sqlite_sequence VALUES('patches',40);
INSERT INTO sqlite_sequence VALUES('config',3);
INSERT INTO sqlite_sequence VALUES('raft_nodes',5);
COMMIT;
```

**Comment by stgraber** [comment]
Quick check list:
 - Check that all nodes are on the same snap revision (`snap list | grep lxd`)
 - Check that none of the nodes have a /var/snap/lxd/common/lxd.debug file in place
 - Confirm that all servers are in sync when it comes to time

**Comment by qskousen** [comment]
All are on 15564.
None have the debug file.
All times seem correct.

I did just notice that the node that has `'config',3` has an external IP for `core.https_address` and an internal one for `cluster.https_address`, would that be a problem? The other nodes have both set to internal.

**Comment by qskousen** [comment]
It is also occasionally saying this: 
```
Jun 16 19:39:31 monkeymisc lxd.daemon[2905]: t=2020-06-16T19:39:31+0000 lvl=warn msg="Failed connecting to global database (attempt 930): failed to create dqlite connection: no available dqlite leader server found"
```

**Comment by stgraber** [comment]
The core.https_address vs cluster.https_address shouldn't matter, LXD will only use cluster.https_address for dqlite communication, so as long as that one is correct you should be fine.

**Comment by stgraber** [comment]
So nothing immediately rings a bell with me and @freeekanayaka is off today, so I'd just recommend you run `systemctl reload snap.lxd.daemon` on all nodes and see if that somehow makes them talk to each other again.

**Comment by qskousen** [comment]
@stgraber yeah we've tried that a few times already and no luck. Also it times out unless we `pkill -9 lxd` in another terminal while stopping/restarting via `systemctl`. 

If we run `lxd cluster recover-from-quorum-loss` can this potentially result in container loss?

**Comment by stgraber** [comment]
This command is only suitable if you're down to a single cluster member, all DB records about anything other than the local node will be lost.

**Comment by stgraber** [comment]
Ok, so further debugging would be to do `snap set lxd daemon.debug=true` on all nodes, then `systemctl reload snap.lxd.daemon` on all of them, give them 5min or so, confirm that everything is still broken and then give us /var/snap/lxd/common/lxd/logs/lxd.log from all of them.

**Comment by qskousen** [comment]
[misc.log](https://github.com/lxc/lxd/files/4789019/misc.log)
[web2.log](https://github.com/lxc/lxd/files/4789024/web2.log)
[monkeyweb1-lxd.log](https://github.com/lxc/lxd/files/4789025/monkeyweb1-lxd.log)
[monkeylb-lxd.log](https://github.com/lxc/lxd/files/4789026/monkeylb-lxd.log)

Here are the results.

**Comment by stgraber** [comment]
They really seem quite convinced that there's a version mismatch going on...

**Comment by stgraber** [comment]
`misc.log` is the only one which does not issue the "nodes are behind" message, not sure if it's because it's not a database node or if it's because it's the one that's somehow running behind.

Can you try running `sqlite3 /var/snap/lxd/common/lxd/database/global/db.bin "SELECT * FROM nodes;"` on every node?
I would expect it to only really work on 3 of them but that may give us a better view as to what's going on with the versions.

**Comment by qskousen** [comment]
misc:
```
1|monkeyweb1||192.168.211.119:8443|28|193|2020-06-16 00:20:56.708871558+00:00|0|2
3|monkeyweb2||192.168.159.233:8443|28|193|2020-06-16 00:20:56.709160779+00:00|0|2
4|monkeylb||192.168.220.80:8443|28|193|2020-06-16 00:20:56.709499978+00:00|0|2
5|monkeymisc||192.168.210.116:8443|28|193|2020-06-16 00:20:56.709644728+00:00|0|2
```

lb:
```
1|monkeyweb1||192.168.211.119:8443|28|189|2020-06-07 13:01:53.61555979+00:00|0|2
3|monkeyweb2||192.168.159.233:8443|28|189|2020-06-07 13:01:53.615033959+00:00|0|2
4|monkeylb||192.168.220.80:8443|28|189|2020-06-07 13:01:53.615277169+00:00|0|2
5|monkeymisc||192.168.210.116:8443|28|189|2020-06-07 13:01:53.615423749+00:00|0|2
```

web1:
```
1|monkeyweb1||192.168.211.119:8443|26|171|2020-03-18 19:54:11.708068348+00:00|0|2
3|monkeyweb2||192.168.159.233:8443|26|171|2020-03-18 19:54:11.708265506+00:00|0|2
4|monkeylb||192.168.220.80:8443|26|171|2020-03-18 19:54:11.708419823+00:00|0|2
5|monkeymisc||192.168.210.116:8443|26|171|2020-03-18 19:54:11.708567112+00:00|0|2
```

web: 
```
1|monkeyweb1||192.168.211.119:8443|29|193|2020-06-16 11:55:45.014681237+00:00|0|2
3|monkeyweb2||192.168.159.233:8443|29|193|2020-06-16 11:55:45.014840007+00:00|0|2
4|monkeylb||192.168.220.80:8443|29|193|2020-06-16 11:55:45.014165238+00:00|0|2
5|monkeymisc||192.168.210.116:8443|29|193|2020-06-16 11:55:45.014472238+00:00|0|2
```

It worked on all of them.

**Comment by stgraber** [comment]
Can you show `readlink -f /proc/$(cat /var/snap/lxd/common/lxd.pid)/exe`?

**Comment by stgraber** [comment]
(On all 4 of them)

**Comment by qskousen** [comment]
They are all pointing at `/snap/lxd/15564/bin/lxd`, which I imagine is correct.

**Comment by stgraber** [comment]
Yep, that's correct and "29|193" as listed in the dump above is the correct version.
Can you show `sqlite3 /var/snap/lxd/common/lxd/database/local.db "SELECT * FROM raft_nodes;"` on all of them?

**Comment by qskousen** [comment]
it's not a problem that only one of them says `29|193`?

`sqlite3 /var/snap/lxd/common/lxd/database/local.db "SELECT * FROM raft_nodes;"`

misc:
```
1|192.168.211.119:8443|2
3|192.168.159.233:8443|2
4|192.168.220.80:8443|2
5|192.168.210.116:8443|0
```

lb:
```
1|192.168.211.119:8443|2
3|192.168.159.233:8443|2
4|192.168.220.80:8443|2
5|192.168.210.116:8443|0
```

web1:
```
1|192.168.211.119:8443|2
3|192.168.159.233:8443|2
4|192.168.220.80:8443|2
5|192.168.210.116:8443|0
```

web2:
```
1|192.168.211.119:8443|2
3|192.168.159.233:8443|2
4|192.168.220.80:8443|2
5|192.168.210.116:8443|0
```

**Comment by qskousen** [comment]
IT should be noted that ~ 30 minutes ago we noticed that web1 (192.168.211.119) reported itself as role 0 when every other node had it listed as role 2; thinking that might be the problem, we updated it to role 2 with sqlite3 and restarted everything, but it didn't seem to make a difference.

**Comment by stgraber** [comment]
Hmm, so this looks wrong to me...
Normally on a 4 nodes cluster, I would expect:
 - 3 nodes at role=0
 - 1 node at role=1

role 0 is voting, role 1 is follower, role 2 is idle

**Comment by stgraber** [comment]
Now, just modifying `local.db` is unlikely to get you super far as that's mostly a local cache of what's in raft and it looks like the raft roles don't make much sense here.

**Comment by stgraber** [comment]
What does `lxd cluster list-database` get you on each node?

**Comment by qskousen** [comment]
They all report `192.168.210.116:8443`

**Comment by qskousen** [comment]
(210.116 corresponds to misc, above)

**Comment by stgraber** [comment]
Ok, just checked the code and well, it just queries local.sql so no surprise that it's consistent with earlier output ;)

**Comment by qskousen** [comment]
Could it be a problem that they are reporting misc as the database server, and misc is `28|193` instead of `29|193`?

**Comment by stgraber** [comment]
It kinda feels like you're in a halfway state between running a cluster with a single voting node and running a cluster with 3 voting nodes and a standby... This is quite weird.

Could you make a tarball of `/var/snap/lxd/common/lxd/database` from misc and e-mail me that to stgraber at ubuntu dot com?

That should let me reproduce that issue and try to see if there's an easy way out of it which doesn't involve waiting for @freeekanayaka to start working tomorrow.

**Comment by stgraber** [comment]
If the role re-shuffle is very recent, I may be able to just discard a recent transaction and have that DB come up, if not, I may be able to force it to behave as if it's in the configuration you see above with only misc being a voting node, then once it's up we can worry about fixing up the roles.

**Comment by qskousen** [comment]
Sent the database.

**Comment by stgraber** [comment]
Okay, so there's something we can try, I'm just hoping that @freeekanayaka can then find a way to get you back to a reasonable state from that.

 - On all nodes, make a clean tarball of `/var/snap/lxd/common/lxd/database`, this will be critical should anything go wrong
 - Stop LXD on all nodes with:
   - echo shutdown > /var/snap/lxd/common/state
   - systemctl stop snap.lxd.daemon.service
 - On misc, do `lxd cluster recover-from-quorum-loss` and then start it with `systemctl start snap.lxd.daemon`
 - Confirm that misc seems to be working properly, if only a bit confused by the others being offlined
 - On the others, do `mv /var/snap/lxd/common/lxd/database/global /var/snap/lxd/common/lxd/database/global.moved`
 - Start them back up too with `systemctl start snap.lxd.daemon`

In that config, `misc` will be the only raft node but that's what your config was showing already, so hopefully things will work well enough for now. In simulation here, I did manage to get two nodes talking so I'm hopeful this will work.

@freeekanayaka will then possibly need to give you a custom `lxd` binary or extra instructions to get back to the expected config of 3 voting members and one stand-by.

**Comment by qskousen** [comment]
Took full snapshots of the servers and tried as you suggested. After doing a recover on misc, it said it was starting as a voting node. All LXC commands still hung, and as expected it said it couldn't connect to the other nodes. Moved the global DB on web1 and start LXD there, but it still seems the same:

```
Jun 16 17:51:28 monkeyweb1 lxd.daemon[5542]: t=2020-06-16T17:51:28-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.159.233:8443: dial: Head \"https://192.168.159.233:8443/internal/database\": dial tcp 192.168.159.233:8443: connect: connection refused"
Jun 16 17:51:28 monkeyweb1 lxd.daemon[5542]: t=2020-06-16T17:51:28-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.210.116:8443: dial: 503 Service Unavailable"
Jun 16 17:51:28 monkeyweb1 lxd.daemon[5542]: t=2020-06-16T17:51:28-0600 lvl=dbug msg="Found cert" name=0
Jun 16 17:51:28 monkeyweb1 lxd.daemon[5542]: t=2020-06-16T17:51:28-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.211.119:8443: dial: some nodes are behind this node's version"
Jun 16 17:51:28 monkeyweb1 lxd.daemon[5542]: t=2020-06-16T17:51:28-0600 lvl=warn msg="Dqlite: attempt 0: server 192.168.220.80:8443: dial: Head \"https://192.168.220.80:8443/internal/database\": dial tcp 192.168.220.80:8443: connect: connection refused"
```

Continued moving global database and starting LXD on the others just in case, but same result as before.

**Comment by stgraber** [comment]
Hmm, this is odd, `misc` should definitely have gone online as it did here, unless it's something failing later than I could go here without having your storage pool on my system.

Can you show the `lxd.log` for misc?

**Comment by qskousen** [comment]
```
t=2020-06-16T23:41:52+0000 lvl=info msg="LXD 4.2 is starting in normal mode" path=/var/snap/lxd/common/lxd
t=2020-06-16T23:41:52+0000 lvl=info msg="Kernel uid/gid map:"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - u 0 0 4294967295"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - g 0 0 4294967295"
t=2020-06-16T23:41:52+0000 lvl=info msg="Configured LXD uid/gid map:"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - u 0 1000000 1000000000"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - g 0 1000000 1000000000"
t=2020-06-16T23:41:52+0000 lvl=info msg="Kernel features:"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - netnsid-based network retrieval: no"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - pidfds: no"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - uevent injection: no"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - seccomp listener: no"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - seccomp listener continue syscalls: no"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - unprivileged file capabilities: yes"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - cgroup layout: hybrid"
t=2020-06-16T23:41:52+0000 lvl=warn msg=" - Couldn't find the CGroup memory swap accounting, swap limits will be ignored"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - shiftfs support: disabled"
t=2020-06-16T23:41:52+0000 lvl=info msg="Initializing local database"
t=2020-06-16T23:41:52+0000 lvl=dbug msg="Initializing database gateway"
t=2020-06-16T23:41:52+0000 lvl=dbug msg="Start database node" address=192.168.210.116:8443 id=5 role=voter
t=2020-06-16T23:41:52+0000 lvl=dbug msg="Detected stale unix socket, deleting"
t=2020-06-16T23:41:52+0000 lvl=info msg="Starting cluster handler:"
t=2020-06-16T23:41:52+0000 lvl=info msg="Starting /dev/lxd handler:"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - binding devlxd socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2020-06-16T23:41:52+0000 lvl=info msg="REST API daemon:"
t=2020-06-16T23:41:52+0000 lvl=info msg=" - binding Unix socket" inherited=true socket=/var/snap/lxd/common/lxd/unix.socket
t=2020-06-16T23:41:52+0000 lvl=info msg=" - binding TCP socket" socket=192.168.210.116:8443
t=2020-06-16T23:41:52+0000 lvl=info msg="Initializing global database"
t=2020-06-16T23:41:52+0000 lvl=warn msg="Dqlite: attempt 0: server 192.168.159.233:8443: dial: Head \"https://192.168.159.233:8443/internal/database\": dial tcp 192.168.159.233:8443: connect: connection refused"
t=2020-06-16T23:41:52+0000 lvl=dbug msg="Found cert" name=0
t=2020-06-16T23:41:52+0000 lvl=warn msg="Dqlite: attempt 0: server 192.168.210.116:8443: dial: 503 Service Unavailable"
t=2020-06-16T23:41:52+0000 lvl=warn msg="Dqlite: attempt 0: server 192.168.211.119:8443: dial: Head \"https://192.168.211.119:8443/internal/database\": dial tcp 192.168.211.119:8443: connect: connection refused"
t=2020-06-16T23:41:52+0000 lvl=warn msg="Dqlite: attempt 0: server 192.168.220.80:8443: dial: Head \"https://192.168.220.80:8443/internal/database\": dial tcp 192.168.220.80:8443: connect: connection refused"
t=2020-06-16T23:41:52+0000 lvl=warn msg="Dqlite: attempt 1: server 192.168.159.233:8443: dial: Head \"https://192.168.159.233:8443/internal/database\": dial tcp 192.168.159.233:8443: connect: connection refused"
```
--- snip---
```
t=2020-06-16T23:42:02+0000 lvl=dbug msg="Failed connecting to global database (attempt 0): failed to create dqlite connection: no available dqlite leader server found"
```

**Comment by qskousen** [comment]
Also, I don't know how the recover command is supposed to work, but when I ran it the terminal sat on the command while LXD started up as shown above, and i had to ctrl+c to get back to the command line, and LXD continued running, so starting it with systemctl was impossible as it showed it was running already.

**Comment by qskousen** [comment]
would it be any different to start the recover command on web2 where it shows the `29|193`?

**Comment by stgraber** [comment]
Based on config, the recover must be done on misc or you'll end up with a db that can never connect.

**Comment by stgraber** [comment]
By the sound of it, the issue is that you didn't `lxd cluster recover-from-quorum-loss` didn't run correctly.

Try this on misc:
 - echo shutdown > /var/snap/lxd/common/lxd/state
 - systemctl stop snap.lxd.daemon snap.lxd.daemon.unix.socket
 - ps aux | grep lxd.*logfile  # this should not show a running lxd
 - `lxd cluster recover-from-quorum-loss`

When I did it here, it just showed the scary warning message, then you enter `yes` and it just returns once it's done.
When that's that, running `systemctl start snap.lxd.daemon` should start LXD back up on misc.

**Comment by qskousen** [comment]
That worked! I'm not sure what happened last time but this time it did as you said, and they are all back up and running!

Somewhat disturbing based on what you said before is what `select * from raft_nodes` shows now:

```
3|192.168.159.233:8443|0
4|192.168.220.80:8443|0
5|192.168.210.116:8443|0
```

119 isn't showing at all, despite being connected, and there are no 1's.

**Comment by stgraber** [comment]
Yeah, this will definitely need some manual fixing of some kind...

Can you send me an updated tarball from misc?

**Comment by stgraber** [comment]
I'll send the old and new tarballs to @freeekanayaka so he can investigate and give instructions on how to get things back in order from there.

In your setup, you really ought to have 3 voting members and one standby, quite confused how that wasn't the case in the first place and why a seemingly random bugfix refresh caused things to hang.

**Comment by stgraber** [comment]
@qskousen I'm going offline for the day, if you get around to make a new tarball of misc, can you e-mail it to both:
 - stgraber at ubuntu dot com
 - free dot ekanayaka at canonical dot com

So that @freeekanayaka who's Europe based can look at both the old tarball (which I forwarded) and the new one when he starts his day.

**Comment by qskousen** [comment]
Sent the new DB.

**Comment by freeekanayaka** [comment]
> This command is only suitable if you're down to a single cluster member, all DB records about anything other than the local node will be lost.

FTR, that's not entirely accurate: no db records will be lost, and you can use this command if you remove the `database/global` directory on all other nodes. This will have the effect of picking a node and make it the only voter in the cluster. Other nodes will be able to connect to it to run db queries, and nothing will be deleted from the db.

**Comment by freeekanayaka** [comment]
> > This command is only suitable if you're down to a single cluster member, all DB records about anything other than the local node will be lost.
> 
> FTR, that's not entirely accurate: no db records will be lost, and you can use this command if you remove the `database/global` directory on all other nodes. This will have the effect of picking a node and make it the only voter in the cluster. Other nodes will be able to connect to it to run db queries, and nothing will be deleted from the db.

I've just noticed that you indeed suggested to follow exactly this procedure, so I assume you understand it well.

**Comment by freeekanayaka** [comment]
Regarding the old `database` directory on the `misc` node, it's in a weird state: all servers in the latest raft configuration stored in the raft log have role `2`, which is stand-by, and none of them is voter. That's the reason no leader gets elected and the database does not come online. Also, the address of the bootstrap node (the first node of the cluster) is stored as `0`, which is weird because it normally should be `1`. This will need further investigation, to check what could have led to this state.

**Comment by freeekanayaka** [comment]
> That worked! I'm not sure what happened last time but this time it did as you said, and they are all back up and running!
> 
> Somewhat disturbing based on what you said before is what `select * from raft_nodes` shows now:
> 
> ```
> 3|192.168.159.233:8443|0
> 4|192.168.220.80:8443|0
> 5|192.168.210.116:8443|0
> ```
> 
> 119 isn't showing at all, despite being connected, and there are no 1's.

In theory, this should be something temporary, eventually the fourth node should show up. If it still isn't there, please can you paste or send me the `lxd.log` file of all four nodes?

**Comment by qskousen** [comment]
As of now they are all still the same as before, `.119` is not listed and they are all 0. I am attaching the log files.

[lb.log](https://github.com/lxc/lxd/files/4793349/lb.log)
[misc.log](https://github.com/lxc/lxd/files/4793351/misc.log)
[web1.log](https://github.com/lxc/lxd/files/4793352/web1.log)
[web2.log](https://github.com/lxc/lxd/files/4793353/web2.log)

**Comment by qskousen** [comment]
Reading through the logfiles I see that web1 (which is IP .119) has early on:

```
t=2020-06-16T18:32:42-0600 lvl=dbug msg="Database error: protocol.Error{Code:517, Message:\"database is locked\"}" 
t=2020-06-16T18:32:42-0600 lvl=dbug msg="Retry failed db interaction (database is locked)" 
```

Sounds like that could be a problem. However, `fuser /var/snap/lxd/common/lxd/database/local.db` gives me `7722` and `ps aux | grep 7722` gives me:

```
root      7722  0.7  0.4 2131312 72752 ?       Sl   Jun16   6:42 lxd --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd --debug
```

which looks correct, so I am not sure.

**Comment by stgraber** [comment]
@freeekanayaka

**Comment by freeekanayaka** [comment]
> As of now they are all still the same as before, `.119` is not listed and they are all 0. I am attaching the log files.
> 
> [lb.log](https://github.com/lxc/lxd/files/4793349/lb.log)
> [misc.log](https://github.com/lxc/lxd/files/4793351/misc.log)
> [web1.log](https://github.com/lxc/lxd/files/4793352/web1.log)
> [web2.log](https://github.com/lxc/lxd/files/4793353/web2.log)

Okay, after looking at the logs and at the code, I figure that there is a bug in this case. Should be fixed by #7578, so in theory the missing node should be eventually be added back once you upgrade your cluster to the next LXD version containing this patch.

**Comment by stgraber** [comment]
Ok, assuming that 7578 did the job, closing.
